src/protocols/LoRaWAN/LoRaWAN.cpp

```diff
@@ -6,54 +6,208 @@
   this->subBand = subBand;
-  this->difsSlots = 2;
-  this->backoffMax = 6;
-  this->enableCSMA = false;
   this->dwellTimeEnabledUp = this->dwellTimeUp != 0;
   this->dwellTimeEnabledDn = this->dwellTimeDn != 0;
-  memset(this->availableChannels, 0, sizeof(this->availableChannels));
+  memset(this->channelPlan, 0, sizeof(this->channelPlan));
 }
 
-void LoRaWANNode::setCSMA(uint8_t backoffMax, uint8_t difsSlots, bool enableCSMA) {
-  this->backoffMax = backoffMax;
-  this->difsSlots = difsSlots;
-  this->enableCSMA = enableCSMA;
+#if defined(RADIOLIB_BUILD_ARDUINO)
+int16_t LoRaWANNode::sendReceive(String& strUp, uint8_t fPort, String& strDown, bool isConfirmed, LoRaWANEvent_t* eventUp, LoRaWANEvent_t* eventDown) {
+  int16_t state = RADIOLIB_ERR_UNKNOWN;
+  
+  const char* dataUp = strUp.c_str();
+
+  // build a temporary buffer
+  // LoRaWAN downlinks can have 250 bytes at most with 1 extra byte for NULL
+  size_t lenDown = 0;
+  uint8_t dataDown[251];
+
+  state = this->sendReceive((uint8_t*)dataUp, strlen(dataUp), fPort, dataDown, &lenDown, isConfirmed, eventUp, eventDown);
+
+  if(state == RADIOLIB_ERR_NONE) {
+    // add null terminator
+    dataDown[lenDown] = '\0';
+
+    // initialize Arduino String class
+    strDown = String((char*)dataDown);
+  }
+
+  return(state);
 }
+#endif
 
-void LoRaWANNode::createNonces() {
-  // set the device credentials
-  LoRaWANNode::hton<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_VERSION], RADIOLIB_LORAWAN_NONCES_VERSION_VAL);
-  LoRaWANNode::hton<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_MODE], this->lwMode);
-  LoRaWANNode::hton<uint8_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_CLASS], this->lwClass);
-  LoRaWANNode::hton<uint8_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_PLAN], this->band->bandNum);
-  LoRaWANNode::hton<uint16_t>(&this->bufferNonces[RADIOLIB_LORAWAN_NONCES_CHECKSUM], this->keyCheckSum);
+int16_t LoRaWANNode::sendReceive(uint8_t* dataUp, size_t lenUp, uint8_t fPort, bool isConfirmed, LoRaWANEvent_t* eventUp, LoRaWANEvent_t* eventDown) {
+  int16_t state = RADIOLIB_ERR_UNKNOWN;
+  
+  // build a temporary buffer
+  // LoRaWAN downlinks can have 250 bytes at most with 1 extra byte for NULL
+  size_t lenDown = 0;
+  uint8_t dataDown[251];
+
+  state = this->sendReceive(dataUp, lenUp, fPort, dataDown, &lenDown, isConfirmed, eventUp, eventDown);
+
+  return(state);
+}
+
+int16_t LoRaWANNode::sendReceive(const char* strUp, uint8_t fPort, uint8_t* dataDown, size_t* lenDown, bool isConfirmed, LoRaWANEvent_t* eventUp, LoRaWANEvent_t* eventDown) {
+  int16_t state = RADIOLIB_ERR_UNKNOWN;
+  
+  state = this->sendReceive((uint8_t*)strUp, strlen(strUp), fPort, dataDown, lenDown, isConfirmed, eventUp, eventDown);
+
+  return(state);
+}
+
+int16_t LoRaWANNode::sendReceive(uint8_t* dataUp, size_t lenUp, uint8_t fPort, uint8_t* dataDown, size_t* lenDown, bool isConfirmed, LoRaWANEvent_t* eventUp, LoRaWANEvent_t* eventDown) {
+  int16_t state = RADIOLIB_ERR_UNKNOWN;
```

Comment:
This is a public method, we need to assume the user may pass null pointers for `dataUp`, `dataDown` and/or `lenDown`. Are those a valid use case? If not, we need to catch that. If yes, we need to catch cases when the pointer is null, but length is non-zero.

Reply:
What would be the appropriate error for this? `RADIOLIB_ERR_MEMORY_ALLOCATION_FAILED`?